docs/sovereign/from-sovereign.md

```diff
@@ -0,0 +1,164 @@
+- Fungible Tokens
+- (Dynamic) Non-Fungible Tokens
+- (Dynamic) Semi-Fungible Tokens
+- (Dynamic) Meta ESDT Tokens
+
+When making the deposit, the user specifies:
+1. A destination address on the Sovereign Chain
+2. `TransferData` if the execution contains a smart contract call, which contains gas, function and arguments
+
+For each deposit in the ESDT-Safe smart contract inside a Sovereign Chain, an outgoing operation will be sent to the bridge service that calls the Header-Verifier and ESDT-Safe contracts in the MultiversX MainChain.
+
+:::note
+Here is the [link](https://github.com/multiversx/mx-sovereign-sc/blob/main/esdt-safe/src/to_sovereign/create_tx.rs) to the `deposit` endpoint
+:::
+
+Each action that can be executed remotely through this contract is called an *Operation*. The endpoint responsible for executing those operations is called `execute_operations`.
+
+#### Execution from inside the Sovereign Chain to the MainChain flow
+1. User deposits token to the ESDT-Safe smart contract on Sovereign.
+2. Outgoing *Operations* are created at the end of the round.
+3. Validators sign all the outgoing *Operations*.
+4. Leader sends *Operations* to the bridge service.
+5. Bridge service sends the *Operations* to the Header-Verifier for registration and verification, and then to ESDT-Safe for execution.
+6. At the end of the execution success/fail, a confirmation event will be added which will be received in sovereign through the observer and then the cross chain transfer will be completed.
+
+### Header-Verifier SC
+
+As mentioned in the [Introduction](cross-chain-execution.md) the Header-Verifier smart contract is to verify signatures, store the *BLS Keys* of the validators and register incoming *Operations*.
+
+```rust
+    #[endpoint(registerBridgeOps)]
+    fn register_bridge_operations(
+        &self,
+        signature: BlsSignature<Self::Api>,
+        bridge_operations_hash: ManagedBuffer,
+        operations_hashes: MultiValueEncoded<ManagedBuffer>,
+    )
+```
+
+Any transaction before being executed has to be registered in this smart contract. The reason behind registering any incoming *Operation* is to create a history of all already registered *Operations* and with that an already registered *Operation* doesn't need to be re-registered and re-executed by the validators.
+
+The registering endpoint follows this flow:
+1. If the incoming `bridge_operations_hash` is not in the `hash_of_hashes_history` storage mapper, otherwise, the endpoint will return a panic.
+2. If the hash of all `operations_hashes` and `bridge_operations_hash` are equal, otherwise, the endpoint will return a panic.
+3. All the `operations_hashes` are registered in the smart contract's storage with the `OperationsHashStatus::NotLocked` status.
+4. The `bridge_operations_hash` is inserted in the `hash_of_hashes_history` storage mapper.
```

Comment:
```suggestion
The registering endpoint operates as follows:
1. Verifies that  `bridge_operations_hash` is not found in the `hash_of_hashes_history` storage mapper, otherwise it will return an error.
2. Verifies that the hash of all `operations_hashes` matches the `bridge_operations_hash, otherwise, the endpoint will return an error.
3. All `operations_hashes` are stored in the smart contract's storage with the status OperationsHashStatus::NotLocked.
4. The `bridge_operations_hash` is added to the `hash_of_hashes_history` storage mapper.
```